File: app/src/main/java/com/example/taxconnect/OrderHistoryActivity.java
package com.example.taxconnect;

import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.taxconnect.adapter.TransactionAdapter;
import com.example.taxconnect.model.TransactionModel;
import com.example.taxconnect.model.UserModel;
import com.google.firebase.auth.FirebaseAuth;

import java.util.List;

public class OrderHistoryActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private TransactionAdapter adapter;
    private ProgressBar progressBar;
    private TextView tvEmpty;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_order_history);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
        }
        toolbar.setNavigationOnClickListener(v -> onBackPressed());

        recyclerView = findViewById(R.id.recyclerView);
        progressBar = findViewById(R.id.progressBar);
        tvEmpty = findViewById(R.id.tvEmpty);

        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        adapter = new TransactionAdapter();
        recyclerView.setAdapter(adapter);

        loadTransactions();
    }

    private void loadTransactions() {
        String uid = FirebaseAuth.getInstance().getUid();
        if (uid == null) return;

        progressBar.setVisibility(View.VISIBLE);

        // First we need to know if the user is a CA or Customer to call the right method
        DataRepository.getInstance().fetchUser(uid, new DataRepository.DataCallback<UserModel>() {
            @Override
            public void onSuccess(UserModel user) {
                if (user == null) {
                    progressBar.setVisibility(View.GONE);
                    return;
                }

                if ("CA".equals(user.getRole())) {
                    fetchCaTransactions(uid);
                } else {
                    fetchCustomerTransactions(uid);
                }
            }

            @Override
            public void onError(String error) {
                progressBar.setVisibility(View.GONE);
                Toast.makeText(OrderHistoryActivity.this, "Error loading profile: " + error, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void fetchCustomerTransactions(String uid) {
        DataRepository.getInstance().getUserTransactions(uid, new DataRepository.DataCallback<List<TransactionModel>>() {
            @Override
            public void onSuccess(List<TransactionModel> data) {
                progressBar.setVisibility(View.GONE);
                if (data == null || data.isEmpty()) {
                    tvEmpty.setVisibility(View.VISIBLE);
                } else {
                    tvEmpty.setVisibility(View.GONE);
                    adapter.setTransactions(data);
                }
            }

            @Override
            public void onError(String error) {
                progressBar.setVisibility(View.GONE);
                Toast.makeText(OrderHistoryActivity.this, "Error: " + error, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void fetchCaTransactions(String uid) {
        DataRepository.getInstance().getCaTransactions(uid, new DataRepository.DataCallback<List<TransactionModel>>() {
            @Override
            public void onSuccess(List<TransactionModel> data) {
                progressBar.setVisibility(View.GONE);
                if (data == null || data.isEmpty()) {
                    tvEmpty.setVisibility(View.VISIBLE);
                } else {
                    tvEmpty.setVisibility(View.GONE);
                    adapter.setTransactions(data);
                }
            }

            @Override
            public void onError(String error) {
                progressBar.setVisibility(View.GONE);
                Toast.makeText(OrderHistoryActivity.this, "Error: " + error, Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: tax/TAX-connect-prototype/tax/app/src/main/java/com/example/taxconnect/features/booking/OrderHistoryActivity.kt
package com.example.taxconnect.features.booking

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.taxconnect.R
import com.example.taxconnect.features.wallet.TransactionAdapter
import com.example.taxconnect.databinding.ActivityOrderHistoryBinding
import com.example.taxconnect.data.models.TransactionModel
import com.example.taxconnect.data.models.UserModel
import com.example.taxconnect.data.repositories.DataRepository
import com.google.firebase.auth.FirebaseAuth
import com.example.taxconnect.core.base.BaseActivity

class OrderHistoryActivity : BaseActivity<ActivityOrderHistoryBinding>() {

    override val bindingInflater: (LayoutInflater) -> ActivityOrderHistoryBinding = ActivityOrderHistoryBinding::inflate
    private lateinit var adapter: TransactionAdapter

    override fun initViews() {
        setSupportActionBar(binding.toolbar)
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
        }
        
        binding.recyclerView.layoutManager = LinearLayoutManager(this)
        adapter = TransactionAdapter()
        binding.recyclerView.adapter = adapter

        loadTransactions()
    }

    override fun setupListeners() {
        binding.toolbar.setNavigationOnClickListener { onBackPressed() }
    }

    override fun observeViewModel() {
        // No ViewModel yet
    }

    private fun loadTransactions() {
        val uid = FirebaseAuth.getInstance().uid ?: return

        binding.progressBar.visibility = View.VISIBLE

        // First we need to know if the user is a CA or Customer to call the right method
        DataRepository.getInstance().fetchUser(uid, object : DataRepository.DataCallback<UserModel> {
            override fun onSuccess(data: UserModel?) {
                if (data == null) {
                    binding.progressBar.visibility = View.GONE
                    return
                }

                if ("CA" == data.role) {
                    fetchCaTransactions(uid)
                } else {
                    fetchCustomerTransactions(uid)
                }
            }

            override fun onError(error: String?) {
                binding.progressBar.visibility = View.GONE
                showToast("Error loading profile: $error")
            }
        })
    }

    private fun fetchCustomerTransactions(uid: String) {
        DataRepository.getInstance().getTransactions(uid, object : DataRepository.DataCallback<List<TransactionModel>> {
            override fun onSuccess(data: List<TransactionModel>?) {
                binding.progressBar.visibility = View.GONE
                if (data == null || data.isEmpty()) {
                    binding.tvEmpty.visibility = View.VISIBLE
                } else {
                    binding.tvEmpty.visibility = View.GONE
                    adapter.setTransactions(data)
                }
            }

            override fun onError(error: String?) {
                binding.progressBar.visibility = View.GONE
                showToast(getString(R.string.error_with_message, error))
            }
        })
    }

    private fun fetchCaTransactions(uid: String) {
        DataRepository.getInstance().getCaTransactions(uid, object : DataRepository.DataCallback<List<TransactionModel>> {
            override fun onSuccess(data: List<TransactionModel>?) {
                binding.progressBar.visibility = View.GONE
                if (data == null || data.isEmpty()) {
                    binding.tvEmpty.visibility = View.VISIBLE
                } else {
                    binding.tvEmpty.visibility = View.GONE
                    adapter.setTransactions(data)
                }
            }

            override fun onError(error: String?) {
                binding.progressBar.visibility = View.GONE
                showToast(getString(R.string.error_with_message, error))
            }
        })
    }
}
